File: Mezilo blockchain/Backend/Frontend/Src/Utils/Blockchain.ts
// frontend/src/utils/blockchain.ts

import * as SHA256 from 'crypto-js/sha256';

export function calculateHash(block: any) {
  const blockData = JSON.stringify(block);
  return SHA256(blockData).toString();
}

export function isValidBlock(newBlock: any, previousBlock: any) {
  // Check if the block's previous hash matches the last block's hash
  if (newBlock.previous_hash !== previousBlock.hash) {
    return false;
  }

  // Validate the block's hash (you can adjust the difficulty here if needed)
  if (calculateHash(newBlock) !== newBlock.hash) {
    return false;
  }

  return true;
}

export function createNewBlock(previousBlock: any, transactions: any[], nonce: number) {
  const index = previousBlock.index + 1;
  const timestamp = Date.now();
  const hash = calculateHash({ index, previous_hash: previousBlock.hash, timestamp, transactions, nonce });

  return {
    index,
    previous_hash: previousBlock.hash,
    timestamp,
    transactions,
    nonce,
    hash
  };
}

File: Mezilo blockchain/Backend/App/Blockchain.py
import hashlib
import json
from time import time


class Block:
    def __init__(self, index, previous_hash, timestamp, transactions, nonce=0):
        self.index = index
        self.previous_hash = previous_hash
        self.timestamp = timestamp
        self.transactions = transactions
        self.nonce = nonce
        self.hash = self.calculate_hash()

    def calculate_hash(self):
        block_data = json.dumps(self.__dict__, sort_keys=True).encode()
        return hashlib.sha256(block_data).hexdigest()


class Blockchain:
    def __init__(self):
        self.chain = []
        self.current_transactions = []
        self.create_block(previous_hash='1', nonce=100)

    def create_block(self, previous_hash, nonce):
        block = Block(
            index=len(self.chain) + 1,
            previous_hash=previous_hash,
            timestamp=time(),
            transactions=self.current_transactions,
            nonce=nonce
        )
        self.current_transactions = []
        self.chain.append(block)
        return block

    def add_transaction(self, sender, recipient, amount):
        self.current_transactions.append({
            'sender': sender,
            'recipient': recipient,
            'amount': amount
        })
        return self.chain[-1].index + 1  # Next block's index

    def get_last_block(self):
        return self.chain[-1] if self.chain else None



def proof_of_work(block, difficulty=4):
    target = '0' * difficulty
    while not block.hash.startswith(target):
        block.nonce += 1
        block.hash = block.calculate_hash()
    return block
